### PR TITLE
Lang item

### LINKED ISSUE #32033: Lang item invalid code compiles
The following code compiles whereas it shouldn't:

```Rust
#![feature(lang_items)]

#[lang = "cookie"]
fn cookie() -> ! {
    loop {}
}

#[lang = "cookie"]
fn cookie2() -> ! {
//~^ ERROR: Duplicate lang item found: `cookie`.
//~| NOTE First definition of `cookie` here.
    loop {}
}

fn main() {}
```

cc @brson 
#31957

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #32033

### CODE REVIEW

File: src/librustc/middle/lang_items.rs [hunk 1]
```diff
@@ -158,6 +158,12 @@ impl<'a, 'v, 'tcx> Visitor<'v> for LanguageItemCollector<'a, 'tcx> {
 
             if let Some(item_index) = item_index {
                 self.collect_item(item_index, self.ast_map.local_def_id(item.id))
+            } else {
+                let item_def_id = self.ast_map.local_def_id(item.id);
+                let span = self.ast_map.span_if_local(item_def_id).unwrap();
+                span_err!(self.session, span, E0522,
+                          "creating new item lang is forbidden: `{}`.",
```
Comment (nagisa) [inline src/librustc/middle/lang_items.rs]: Why not just, uh, “definition of a unknown language item”?

Comment (GuillaumeGomez) [inline src/librustc/middle/lang_items.rs]: Way better.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -1962,6 +1962,19 @@ each method; it is not possible to annotate the entire impl with an `#[inline]`
 attribute.
 "##,
 
+E0522: r##"
+A new lang item was created in user code, which is forbidden. Erroneous code example:
```
Comment (nikomatsakis) [inline src/librustc/diagnostics.rs]: Perhaps: "The lang attribute is intended for marking special items that are built-in to Rust itself. This includes special traits (like `Copy` and `Sized`) that affect how the compiler behaves, as well as special functions that may be automatically invoked (such as the handler for out-of-bounds accesses when indexing a slice)."

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Great one, thanks!

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -1962,6 +1962,23 @@ each method; it is not possible to annotate the entire impl with an `#[inline]`
 attribute.
 "##,
 
+E0522: r##"
+The lang attribute is intended for marking special items that are built-in to
+Rust itself. This includes special traits (like `Copy` and `Sized`) that affect
+how the compiler behaves, as well as special functions that may be automatically
+invoked (such as the handker for out-of-bounds accesses when indexing a slice).
```
Comment (nikomatsakis) [inline src/librustc/diagnostics.rs]: Nit: s/handker/handler/

File: src/librustc/middle/lang_items.rs [hunk 4]
```diff
@@ -158,6 +158,12 @@ impl<'a, 'v, 'tcx> Visitor<'v> for LanguageItemCollector<'a, 'tcx> {
 
             if let Some(item_index) = item_index {
                 self.collect_item(item_index, self.ast_map.local_def_id(item.id))
+            } else {
+                let item_def_id = self.ast_map.local_def_id(item.id);
+                let span = self.ast_map.span_if_local(item_def_id).unwrap();
```
Comment (nikomatsakis) [inline src/librustc/middle/lang_items.rs]: why not call `self.ast_map.span(item.id)`? Then you can skip creating the `item_def_id` too.

Comment (GuillaumeGomez) [inline src/librustc/middle/lang_items.rs]: Indeed, it's fixed.

**Comment by GuillaumeGomez** [comment]
Updated.

**Pull request merged**